p2p/peer.go

```diff
@@ -213,17 +217,58 @@ func newPeer(
 		metrics:       NopMetrics(),
 		mlc:           mlc,
 		traceClient:   trace.NoOpTracer(),
+		streams:       make(map[byte]quic.Stream),
 	}
 
-	p.mconn = createMConnection(
-		pc.conn,
-		p,
-		reactorsByCh,
-		msgTypeByChID,
-		chDescs,
-		onPeerError,
-		mConfig,
-	)
+	p.onReceive = func(chID byte, msgBytes []byte) {
+		reactor := reactorsByCh[chID]
+		if reactor == nil {
+			// Note that its ok to panic here as it's caught in the conn._recover,
+			// which does onPeerError.
+			panic(fmt.Sprintf("Unknown channel %X", chID))
+		}
+		mt := msgTypeByChID[chID]
+		msg := proto.Clone(mt)
+		err := proto.Unmarshal(msgBytes, msg)
+		if err != nil {
+			p.Logger.Error("before panic", "msg", msg, "channel", chID, "type", mt, "bytes", hex.EncodeToString(msgBytes))
+			return
+		}
+
+		if w, ok := msg.(Unwrapper); ok {
+			msg, err = w.Unwrap()
+			if err != nil {
+				panic(fmt.Errorf("unwrapping message: %s", err))
+			}
+		}
+
+		labels := []string{
+			"peer_id", string(p.ID()),
+			"chID", fmt.Sprintf("%#x", chID),
+		}
+
+		p.metrics.PeerReceiveBytesTotal.With(labels...).Add(float64(len(msgBytes)))
+		p.metrics.MessageReceiveBytesTotal.With(append(labels, "message_type", p.mlc.ValueToMetricLabel(msg))...).Add(float64(len(msgBytes)))
+		schema.WriteReceivedBytes(p.traceClient, string(p.ID()), chID, len(msgBytes))
+		if nr, ok := reactor.(EnvelopeReceiver); ok {
+			nr.ReceiveEnvelope(Envelope{
+				ChannelID: chID,
+				Src:       p,
+				Message:   msg,
+			})
+		} else {
+			reactor.Receive(chID, p, msgBytes)
+		}
+	}
+
+	go func() {
+		err := p.StartReceiving()
+		if err != nil {
+			p.Logger.Error("error when receiving data from peer", "err", err.Error())
+			onPeerError(p, err)
+		}
+	}()
```

Comment:
This `StartReceiving` is called here in `onReceive` func, while the `onReceive` itself is called by `StartReceiving`. There is a deadloop.

Reply:
`StartReceiving` is not called inside `onReceive` if I see correctly

Reply:
You are right. The diff confused me and it looked the selected code block is inside the onReceive